### ISSUE TITLE
Neovim badly interacts with server when `position_encoding` is set to "utf-8"

### ISSUE THREAD

**Issue by fit-sizhe** [Myriad-Dreamin/tinymist#766]
**Describe the bug**
When typing Chinese in neovim, syntax highlighting is off, and unclosed delimiter errors are everywhere 
![image](https://github.com/user-attachments/assets/a9322d4e-cd7b-4c77-b343-79884dffd000)

**Package/Software version:**
tinymist version: `v0.12.0`, installed through `mason.nvim`
neovim version: 
```
NVIM v0.10.2
Build type: RelWithDebInfo
LuaJIT 2.1.1713484068
```

**Comment by fit-sizhe** [comment]
my lsp settings for tinymist is mentioned in #761.

**Comment by oakrotka** [comment]
I'm also experiencing this with polish characters. My lsp settings:
```lua

lspconfig.tinymist.setup {
    on_attach = on_attach,
    capabilities = capabilities,
    offset_encoding = "utf-8",
    settings = {
      semantic_tokens = "disable",
    }
}
```

**Comment by random-username-here** [comment]
Got a similar error when using cyrrillic letters in Neovim.

The error can be reproduced like this:
 1. Open a file
 2. Type `**`
 3. Move cursor to the middle, write `аа` (cyrillic letter `A` two times)
 4. Get that "unclosed delimiter" error

When looking through the logs, I see this:

```
[2024-11-09T15:38:57Z INFO  sync_lsp] notifing textDocument/didChange succeeded in 152.43µs
[2024-11-09T15:38:57Z INFO  tinymist::actor::editor] received status request(primary) Compiling
error: unclosed delimiter
  ┌─ /home/i-s-d/homework/test.typ:1:0
  │
1 │ *а*а
  │ ^
[2024-11-09T15:38:57Z INFO  tinymist::actor::editor] received diagnostics from primary:9: diag(Some(1))
2024-11-09T15:38:57Z INFO  tinymist::actor::editor] received status request(primary) CompileError
[2024-11-09T15:38:57Z INFO  tinymist::task::cache] CacheEvictTask: evict cache in 8.48µs
```

Which probably means there is something wrong with incremental document changes.

If I type `аааа` instead of `аа`, it also has that error, but it also panics:

```
[ERROR][2024-11-09 18:38:57] .../vim/lsp/rpc.lua:770    "rpc"   "/usr/bin/tinymist"     "stderr"
  [2024-11-09T15:38:57Z INFO  sync_lsp] notifying textDocument/didChange - at Instant { tv_sec: 27190, tv_nsec: 960419617 }
  thread 'main' panicked at crates/tinymist/src/server.rs:962:26:
  invalid range
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[ERROR][2024-11-09 18:39:04] .../vim/lsp/rpc.lua:770    "rpc"   "/usr/bin/tinymist"     "stderr"
  [2024-11-09T15:39:04Z INFO  sync_lsp] handling shutdown - (7) at Instant { tv_sec: 27197, tv_nsec: 433124982 }
  [2024-11-09T15:39:04Z INFO  sync_lsp] handled  shutdown - (7) in 51.50µs
```
(I formatted the LSP log a little bit, because it joined multiple server logs into LSP client log)

My current LSP config:
```lua
lspconfig.tinymist.setup {
  capabilities = capabilities,
  offset_encoding = "utf-8",
  setup = {
    cmd = { 'tinymist' }
  },
  settings = {
    semantic_tokens = "disable",
  }
}
```

I tried setting `RUST_BACKTRACE=1` in the `cmd`, but seems like it is ignored somehow.

**Comment by Eric-Song-Nop** [comment]
A temporary solution might be to upgrade or downgrade your Neovim, I personally think supporting only utf-16 from Tinymist is better.

**Comment by fit-sizhe** [comment]
@Eric-Song-Nop I don't think upgrading will help much, as my neovim version is the latest.

**Comment by Eric-Song-Nop** [comment]
The requirement for utf-8 offset_encoding should be fixed in https://github.com/neovim/neovim/pull/30747 , after that pr I hope that not using utf-8 offset_encoding should be enough.

**Comment by fit-sizhe** [comment]
@Eric-Song-Nop Nope, removing "offset_encoding" causes the issue #761.

**Comment by Myriad-Dreamin** [comment]
I wouldn't like to make an assertion but there should be some bug(s) on neovim and a way without having to change tinymist to make semantic highlighting perfect on neovim, because we haven't received any confused semantic highlighted errors on VS Code for months.

I also add another recent finding about it, hinting there might be inconsistent uses of position encoding on some distros of neovim, which causes desync state between LSP server (tinymist) and clients (neovim). This happens even if tinymist and neovim have negotiated a position encoding: https://github.com/folke/trouble.nvim/issues/606

More information is needed as I'm not a proficient neovim user.

**Comment by SylvanFranklin** [comment]
I agree with @Myriad-Dreamin, this seems like more of a neovim issue than Tinymist. I'm no longer experiencing this issue on the latest neovim and tinymist.
